test/test_tcti.py

```diff
@@ -73,6 +104,103 @@ def test_tctildr(self):
         with self.assertRaises(TypeError):
             TCTILdr(name=1234, conf=None)
 
+    def test_custom_pytcti_esapi(self):
+
+        t = MyTCTI(self.tcti)
+        e = ESAPI(t)
+        e.get_random(4)
+
+        e.startup(TPM2_SU.CLEAR)
+
+    def test_custom_pytcti_C_wrapper_transmit_receive(self):
+
+        t = MyTCTI(self.tcti)
+
+        # Go through the C API directly and call transmit and recv
+        t.transmit(b"\x80\x01\x00\x00\x00\x0C\x00\x00\x01\x44\x00\x00")
+        resp = t.receive(-1)
+        self.assertEqual(resp, b"\x80\x01\x00\x00\x00\n\x00\x00\x01\x00")
+
+    def test_custom_pytcti_cancel(self):
+        if getattr(self.tcti, "name", "") == "swtpm":
+            self.skipTest("cancel not supported by swtpm")
+
+        t = MyTCTI(self.tcti)
+
+        t.transmit(b"\x80\x01\x00\x00\x00\x0C\x00\x00\x01\x44\x00\x00")
+        t.cancel()
+
+    def test_custom_pytcti_finalize(self):
+        # This is NOP for the PyTCTI but nothing should break
+        # TODO maybe it should implement it?
+        t = MyTCTI(self.tcti)
+        t.finalize()
+
```

Comment:
Would be good to have finalize implemented, it's a clear signal from the caller that the instance is not needed anymore

Reply:
> Would be good to have finalize implemented, it's a clear signal from the caller that the instance is not needed anymore

My thought that `__del__` on the Python object was the clear signal, no?

I also added a test for the context manager and just mocked finalize to ensure that it is called. It just NOPs in the Python TCTI Implementation as the Finalize Function Pointer is NULL.

My inclination is to not implement it, but it's always easy to add if someone bumps into it... @whooo play devils advocate here, I could be convinced to add it now with a good argument.

Reply:
I'll take any changes as a next patchset in the series... merged now, but lets keep the discussion going.

Reply:
> The issue with `__del__` is that you don't know when it's actually called, as that is up to the garbage collector, so finalize would be more like `__exit__` or ESAPI.close. There is also the case where you use a PyTCTI subclass as a proxy TCTI, where you might want to proxy finalize. An implementation could of course ignore it by using a default method from PyTCTI.

Ahh yeah consider you have a TCTI that blocks other users from the resource, you want an explicitly timed event so others can establish a TCTI connection. Sold.